File: src/_cpp/TowerDefenseBase/game/unit/Skills.cpp
//
//  Skills.cpp
//  IslandDefense
//
//  Created by Vladimir Tolmachev on 26.11.14.
//
//

#include "Skills.h"
#include "Unit.h"
#include "GameGS.h"
#include "gameboard.h"
#include "consts.h"

NS_CC_BEGIN;

/******************************************************************************/
//MARK: UnitSkill
/******************************************************************************/
UnitSkill::UnitSkill()
: _unit(nullptr)
, _needUnitSkillLevel(0)
{}

UnitSkill::~UnitSkill()
{}

bool UnitSkill::init( const pugi::xml_node & xmlNode, Unit* unit )
{
	_unit = unit;
	_onlyState = xmlNode.attribute( "onlystate" ).as_string();
	_needUnitSkill = xmlNode.attribute( "unitskill" ).as_string();
	_needUnitSkillLevel = xmlNode.attribute( "unitskilllevel" ).as_int();
	return _unit != nullptr;
}

/******************************************************************************/
//MARK: UnitSkillMedic
/******************************************************************************/
UnitSkillMedic::UnitSkillMedic()
: _radius( 0 )
, _frequence( 0 )
, _health( 0 )
, _duration( 0 )
, _timer( 0 )
, _timerDuration( 0 )
, _effectDescription()
, _maxTargets( -1 )
, _execution(false)
{}

UnitSkillMedic::~UnitSkillMedic()
{}

bool UnitSkillMedic::init( const pugi::xml_node & xmlNode, Unit* unit )
{
	UnitSkill::init( xmlNode, unit );

	_radius = xmlNode.attribute( "radius" ).as_float( 0.f );
	_frequence = xmlNode.attribute( "frequence" ).as_float( 0.f );
	_health = xmlNode.attribute( "health" ).as_float( 0.f );
	_duration = xmlNode.attribute( "duration" ).as_float( 0.f );
	_effectDescription = xmlNode.attribute( "effect_description" ).as_string();
	_maxTargets = xmlNode.attribute( "maxtargets" ).as_int();

	std::list<std::string> units;
	split( units, xmlNode.attribute( "units" ).as_string() );
	for( auto unit : units )
		_allowUnits.insert( unit );

	return true;
}

void UnitSkillMedic::update( float dt, Unit* context )
{
	assert( context );

	_timer += dt;
	if( _timer > _frequence )
	{
		if( _timerDuration < 0.001f )
		{
			execute( context );
		}
		if( _timerDuration >= _duration )
		{
			stop( context );
			_timer = 0;
			_timerDuration = 0;
		}
		else
		{
			_timerDuration += dt;
		}
	}
}

bool UnitSkillMedic::execution()
{
	return _execution;
}

void UnitSkillMedic::execute( Unit* context )
{
	assert( GameGS::getInstance() );
	const auto& board = GameGS::getInstance()->getGameBoard();
	std::vector<Unit*> targets;
	Point center = context->getPosition();
	board.getTargetsByRadius( targets, center, _radius );

	unsigned counter = _maxTargets;
	bool wasHealing( false );
	for( auto target : targets )
	{
		if( _allowUnits.empty() == false )
		{
			bool allow = _allowUnits.find( target->getName() ) != _allowUnits.end();
			if( allow == false )
				continue;
		}
		if( target->getType() != UnitType::creep )
			continue;

		float health = target->getCurrentHealth();
		float max = target->getHealth();

		float value = std::min( max, health + _health );
		if( health < max )
		{
			target->setCurrentHealth( value );
			GameGS::getInstance()->createEffect( context, target, _effectDescription );
			wasHealing = true;
			--counter;
		}
		if( counter == 0 )
			break;
	}

	if( wasHealing )
	{
		_execution = true;
		context->stop();
		context->runEvent( "on_healing" );
	}
}

void UnitSkillMedic::stop( Unit* context )
{
	_execution = false;
	context->move();
}


/******************************************************************************/
//MARK: UnitSkillRunTasksByTime
/******************************************************************************/
UnitSkillRunTasksByTime::UnitSkillRunTasksByTime()
: _timer( 0 )
, _timerDuration(0)
, _frequence( 0 )
, _events()
, _stopedUnit(false)
, _resumeMoving(false)
, _stopDuration(0)
, _count(-1)
{}

UnitSkillRunTasksByTime::~UnitSkillRunTasksByTime()
{}

bool UnitSkillRunTasksByTime::init( const pugi::xml_node & xmlNode, Unit* unit )
{
	UnitSkill::init( xmlNode, unit );

	_frequence = xmlNode.attribute( "frequence" ).as_float( 0.f );
	_stopedUnit = xmlNode.attribute( "stopunit" ).as_bool( false );
	_stopDuration = xmlNode.attribute( "stopduration" ).as_float( 0.f );
	_count = xmlNode.attribute( "count" ).as_int( -1 );

	auto xmlEvents = xmlNode.child( "eventlist" );
	FOR_EACHXML( xmlEvents, xmlEvent )
	{
		auto event = xmlLoader::load_event( xmlEvent );
		if( event )
			_events.push_back( event );
	}
	return true;
}

void UnitSkillRunTasksByTime::update( float dt, Unit* context )
{
	if( _stopedUnit )
	{
		_timer += dt;
		if( _timer > _frequence )
		{
			if( _timerDuration < 0.001f )
			{
				execute( context );
			}
			if( _timerDuration >= _stopDuration )
			{
				stop( context );
				_timer = 0;
				_timerDuration = 0;
			}
			else
			{
				_timerDuration += dt;
			}
		}
	}
	else
	{
		_timer += dt;
		if( _timer > _frequence )
		{
			_timer -= _frequence;
			execute( context );
			stop( context );
		}
	}
}

void UnitSkillRunTasksByTime::execute( Unit* context )
{
	if( _count > 0 )
	{
		--_count;
		_events.execute( context );
		if( _stopedUnit )
		{
			_resumeMoving = context->current_state().get_name() == MachineUnit::State::state_move;
			if( _resumeMoving )
				context->stop();
		}
	}
}

void UnitSkillRunTasksByTime::stop( Unit* context )
{
	if( _resumeMoving )
		context->move();
}

/*****************************************************************************/
//MARK: UnitSkillCounter
/*****************************************************************************/

UnitSkillCounter::UnitSkillCounter()
: _isActive( false )
, _damageCounter( { false, 0, 0 } )
, _timeCounter( { false, 0, 0 } )
, _damageCounterActive( { false, 0, 0 } )
, _timeCounterActive( { false, 0, 0 } )
{}

UnitSkillCounter::~UnitSkillCounter()
{}

bool UnitSkillCounter::init( const pugi::xml_node & xmlNode, Unit* unit )
{
	do
	{
		CC_BREAK_IF( !UnitSkill::init( xmlNode, unit ) );

		auto xmlDamage = xmlNode.attribute( "damage" );
		auto xmlDamageActive = xmlNode.attribute( "damageactive" );
		auto xmlTime = xmlNode.attribute( "time" );
		auto xmlTimeActive = xmlNode.attribute( "timeactive" );

		if( xmlDamage )
		{
			_damageCounter.active = true;
			_damageCounter.def = _damageCounter.left = xmlDamage.as_int();
		}
		if( xmlDamageActive )
		{
			_damageCounterActive.active = true;
			_damageCounterActive.def = _damageCounterActive.left = xmlDamageActive.as_int();
		}
		if( xmlTime )
		{
			_timeCounter.active = true;
			_timeCounter.def = _timeCounter.left = xmlTime.as_float();
		}
		if( xmlTimeActive )
		{
			_timeCounterActive.active = true;
			_timeCounterActive.def = _timeCounterActive.left = xmlTimeActive.as_float();
		}

		_type = xmlNode.attribute( "skilltype" ).as_string();
		_value = xmlNode.attribute( "skillvalue" ).as_float();

		return true;
	}
	while( false );
	return false;

}

void UnitSkillCounter::update( float dt, Unit* context )
{
	auto& counter = _isActive ? _timeCounterActive : _timeCounter;
	if( counter.action( dt ) )
	{
		if( _isActive ) executeBack();
		else execute();
	}
}

void UnitSkillCounter::onDamage( float damage )
{
	auto& counter = _isActive ? _damageCounterActive : _damageCounter;
	if( counter.action( 1 ) )
	{
		if( _isActive ) executeBack();
		else execute();
	}
}

void UnitSkillCounter::execute()
{
	_isActive = true;
	getUnit()->skillActivated( this );
}

void UnitSkillCounter::executeBack()
{
	_damageCounterActive.left = _damageCounterActive.def;
	_timeCounterActive.left = _timeCounterActive.def;
	_isActive = false;
	getUnit()->skillDeactivated( this );
}


/*****************************************************************************/
//MARK: UnitSkillRateParameter
/*****************************************************************************/
UnitSkillRateParameter::UnitSkillRateParameter() 
: _rate( 1 )
, _parameter()
{}

UnitSkillRateParameter::~UnitSkillRateParameter() 
{}

bool UnitSkillRateParameter::init( const pugi::xml_node & xmlNode, Unit* unit )
{
	UnitSkill::init( xmlNode, unit );

	_rate = xmlNode.attribute( "rate" ).as_float( 1.f );
	_parameter = xmlNode.attribute( "parameter" ).as_string();
	return true;
}

void UnitSkillRateParameter::update( float dt, Unit* context ) {}

NS_CC_END;

File: src/_eu/game/unit/Skills.js
/******************************************************************************
 * Copyright 2014-2016 Vladimir Tolmachev
 * Copyright 2016 Visionarity AG
 * Vladimir Tolmachev and Visionarity AG have unlimited commercial
 * licenses for commercial use and customization
 *
 * Author: Vladimir Tolmachev (tolm_vl@hotmail.com)
 * Ported C++ to Javascript: Visionarity AG / Vladimir Tolmachev
 * Project: Island Defense (JS)
 * If you received the code not from the author, please contact us
 ******************************************************************************/

//Define namespace
var EU = EU || {};

EU.UnitSkill = cc.Class.extend(
{
    //virtual void update( var dt, EU.Unit context )=0;
    /** @type {EU.Unit} */ _unit : null,
    /** @type {String} */ _onlyState : null,
    /** @type {String} */ _needUnitSkill : null,
    /** @type {int} */ _needUnitSkillLevel : 0,

    getNeedUnitSkill: function(){return this._needUnitSkill; },
    getNeedUnitSkillLevel: function(){return this._needUnitSkillLevel; },
    getOnlyState: function(){return this._onlyState;},
    getUnit: function(){return this._unit;},
    onDamage: function(){},
    execution: function () { return false; },



/******************************************************************************/
    //MARK: UnitSkill
    /******************************************************************************/

    ctor: function( /**@type {Element} */ xmlNode, /**@type {EU.Unit} */ unit )
    {
        this._unit = unit;
        this._onlyState = xmlNode.getAttribute( "onlystate" );
        this._needUnitSkill = xmlNode.getAttribute( "unitskill" );
        this._needUnitSkillLevel = EU.Common.strToInt(xmlNode.getAttribute( "unitskilllevel" ));
        if( !this._onlyState )
            this._onlyState = "";
        if( !this._needUnitSkill )
            this._needUnitSkill = "";
        return this._unit != null;
    }

});

EU.UnitSkillMedic = EU.UnitSkill.extend(
{
    /** @type {Number} */ _radius : null,
    /** @type {Number} */ _frequence : null,
    /** @type {Number} */ _health : null,
    /** @type {Number} */ _duration : null,
    /** @type {Number} */ _timer : null,
    /** @type {Number} */ _timerDuration : null,
    /** @type {int} */ _maxTargets : null,
    /** @type {std.string} */ _effectDescription : null,
    /** @type {std.set<std.string>} */ _allowUnits : null,
    /** @type {Boolean} */ _execution : null,


    /******************************************************************************/
    //MARK: UnitSkillMedic
    /******************************************************************************/
    ctor: function( /**@type {Element} */ xmlNode, /**@type {EU.Unit} */ unit )
    {
        "use strict";

        this._radius = 0;
        this._frequence = 0;
        this._health = 0;
        this._duration = 0;
        this._timer = 0;
        this._timerDuration = 0;
        this._effectDescription = "";
        this._maxTargets = -1 ;
        this._allowUnits = [] ;
        this._execution = false ;

        this._super(xmlNode, unit );

        this._radius = EU.Common.strToFloat(xmlNode.getAttribute( "radius" ));
        this._frequence = EU.Common.strToFloat(xmlNode.getAttribute( "frequence" ));
        this._health = EU.Common.strToFloat(xmlNode.getAttribute( "health" ));
        this._duration = EU.Common.strToFloat(xmlNode.getAttribute( "duration" ));
        this._effectDescription = xmlNode.getAttribute( "effect_description" );
        this._maxTargets = EU.Common.strToInt(xmlNode.getAttribute( "maxtargets" ));

        var units = [];
        units = xmlNode.getAttribute( "units").split(',');
        for (var i = 0; i < units.length; i++) {
            var unit = units[i];
            if (!this._allowUnits.indexOf(unit)) this._allowUnits.push( unit );
        }

        return true;
    },

    update: function( /**@type {var} */ dt, /**@type {EU.Unit} */ context )
    {
        EU.assert( context );

        this._timer += dt;
        if( this._timer > this._frequence )
        {
            if( this._timerDuration < 0.001  )
            {
                this.execute( context );
            }
            if( this._timerDuration >= this._duration )
            {
                this.stop( context );
                this._timer = 0;
                this._timerDuration = 0;
            }
            else
            {
                this._timerDuration += dt;
            }
        }
    },

    execution: function()
    {
        return this._execution;
    },

    execute: function( /**@type {EU.Unit} */ context )
    {
        EU.assert( EU.GameGS );
        var board = EU.GameGSInstance.getGameBoard();
        /**
         * @type {Array<EU.Unit>}
         */
        var targets=[];
        var center = context.getPosition();
        board.getTargetsByRadius( targets, center, this._radius );

        var counter = this._maxTargets;
        var wasHealing =false ;
        for (var i = 0; i < targets.length; i++) {
            var target = targets[i];
            if( this._allowUnits.length > 0 )
            {
                var allow = this._allowUnits.indexOf( target.getName() ) >= 0;
                if( allow == false )
                    continue;
            }
            if( target.getType() != EU.UnitType.creep )
                continue;

            var health = target.getCurrentHealth();
            var max = target._health;

            var value = Math.min( max, health + this._health );
            if( health < max )
            {
                target.setCurrentHealth( value );
                EU.GameGSInstance.createEffect( context, target, this._effectDescription );
                wasHealing = true;
                --counter;
            }
            if( counter == 0 )
                break;
        }

        if( wasHealing )
        {
            this._execution = true;
            context.stop();
            context.runEvent( "on_healing" );
        }
    },

    stop: function( /**@type {EU.Unit} */ context )
    {
        this._execution = false;
        context.move();
    }
});

EU.UnitSkillRunTasksByTime = EU.UnitSkill.extend(
{
    /** @type {Number} */ _timer : null,
    /** @type {Number} */ _timerDuration : null,
    /** @type {Number} */ _frequence : null,
    /** @type {Boolean} */ _stopedUnit : null,
    /** @type {Boolean} */ _resumeMoving : null,
    /** @type {Number} */ _stopDuration : null,
    /** @type {int} */ _count : null,
    /** @type {EventsList} */ _events : null,


    /******************************************************************************/
    //MARK: UnitSkillRunTasksByTime
    /******************************************************************************/

    ctor: function( /**@type {Document} */ xmlNode, /**@type {EU.Unit} */ unit )
    {
        this._timer = 0;
        this._timerDuration = 0;
        this._frequence = 0;
        this._events = [];
        this._stopedUnit = false;
        this._resumeMoving = false;
        this._stopDuration = 0;
        this._count = -1;

        this._super(xmlNode, unit);

        this._frequence = EU.Common.strToFloat(xmlNode.getAttribute( "frequence" ));
        this._stopedUnit = EU.Common.strToBool(xmlNode.getAttribute( "stopunit" ));
        this._stopDuration = EU.Common.strToFloat(xmlNode.getAttribute( "stopduration" ) );
        this._count = EU.Common.strToInt(EU.asObject(xmlNode.getAttribute( "count" ),"-1" ));

        var xmlEvents = xmlNode.getElementsByTagName( "eventlist" )[0];
        for (var i = 0; i < xmlEvents.children.length; i++) {
            var xmlEvent = xmlEvents.children[i];
            var event = EU.xmlLoader.load_event_xml_node( xmlEvent );
            if( event )
                this._events.push( event );
        }
        return true;
    },

    update: function( /**@type {var} */ dt, /**@type {EU.Unit} */ context )
    {
        if( this._stopedUnit )
        {
            this._timer += dt;
            if( this._timer > this._frequence )
            {
                if( this._timerDuration < 0.001  )
                {
                    this.execute( context );
                }
                if( this._timerDuration >= this._stopDuration )
                {
                    this.stop( context );
                    this._timer = 0;
                    this._timerDuration = 0;
                }
                else
                {
                    this._timerDuration += dt;
                }
            }
        }
        else
        {
            this._timer += dt;
            if( this._timer > this._frequence )
            {
                this._timer -= this._frequence;
                this.execute( context );
                this.stop( context );
            }
        }
    },

    execute: function( /**@type {EU.Unit} */ context )
    {
        if( this._count > 0 )
        {
            --this._count;
            this._events.execute( context );
            if( this._stopedUnit )
            {
                this._resumeMoving = context.current_state().get_name() == EU.MachineUnit.State.state_move;
                if( this._resumeMoving )
                    context.stop();
            }
        }
    },

    stop: function( /**@type {EU.Unit} */ context )
    {
        if( this._resumeMoving )
            context.move();
    }

});

EU.UnitSkillCounter = EU.UnitSkill.extend(
{
    Counter : function(active, def, left)
    {
        this.active = active;
        this.def = def;
        this.left = left;
        this.self = this;
        this.action = function(decrement)
        {
            if( !self.active ) return false;
            self.left -= decrement;
            if( self.left <= 0 )
            {
                self.left = self.def;
                return true;
            }
            return false;
        }
    },
    /** @type {Boolean} */ _isActive : null,
    /** @type {Counter<int>} */ _damageCounter : null,
    /** @type {Counter<Number>} */ _timeCounter : null,
    /** @type {Counter<int>} */ _damageCounterActive : null,
    /** @type {Counter<Number>} */ _timeCounterActive : null,
    /** @type {Number} */ _value : null,
    /** @type {String} */ _type : null,

    /*****************************************************************************/
    //MARK: UnitSkillCounter
    /*****************************************************************************/
    
    ctor: function( /**@type {Element} */ xmlNode, /**@type {EU.Unit} */ unit )
    {
        this._isActive = false ;
        this._damageCounter = new this.Counter(false, 0, 0  );
        this._timeCounter = new this.Counter(false, 0, 0  );
        this._damageCounterActive = new this.Counter(false, 0, 0  );
        this._timeCounterActive = new this.Counter(false, 0, 0  );

        this._super(xmlNode, unit);

        var xmlDamage = xmlNode.getAttribute( "damage" );
        var xmlDamageActive = xmlNode.getAttribute( "damageactive" );
        var xmlTime = xmlNode.getAttribute( "time" );
        var xmlTimeActive = xmlNode.getAttribute( "timeactive" );

        if( xmlDamage )
        {
            this._damageCounter.active = true;
            this._damageCounter.def = this._damageCounter.left = parseInt(xmlDamage);
        }
        if( xmlDamageActive )
        {
            this._damageCounterActive.active = true;
            this._damageCounterActive.def = this._damageCounterActive.left = parseInt(xmlDamageActive);
        }
        if( xmlTime )
        {
            this._timeCounter.active = true;
            this._timeCounter.def = this._timeCounter.left = parseFloat(xmlTime);
        }
        if( xmlTimeActive )
        {
            this._timeCounterActive.active = true;
            this._timeCounterActive.def = this._timeCounterActive.left = parseFloat(xmlTimeActive);
        }

        this._type = xmlNode.getAttribute( "skilltype" );
        this._value = parseFloat(xmlNode.getAttribute( "skillvalue" ));

        return true;
    },
    
    update: function( /**@type {var} */ dt, /**@type {EU.Unit} */ context )
    {
        var counter = this._isActive ? this._timeCounterActive : this._timeCounter;
        if( counter.action( dt ) )
        {
            if( this._isActive ) this.executeBack();
            else this.execute();
        }
    },
    
    onDamage: function( /**@type {var} */ damage )
    {
        var counter = this._isActive ? this._damageCounterActive : this._damageCounter;
        if( counter.action( 1 ) )
        {
            if( this._isActive ) this.executeBack();
            else this.execute();
        }
    },
    
    execute: function()
    {
        this._isActive = true;
        this.getUnit().skillActivated( this );
    },
    
    executeBack: function()
    {
        this._damageCounterActive.left = this._damageCounterActive.def;
        this._timeCounterActive.left = this._timeCounterActive.def;
        this._isActive = false;
        this.getUnit().skillDeactivated( this );
    }

});

EU.UnitSkillRateParameter = EU.UnitSkill.extend(
{
    /** @type {Number} */ _rate : null,
    /** @type {String} */ _parameter : null,

    getParameter: function(){ return this._parameter;},
    getRate: function(){ return this._rate;},

    /*****************************************************************************/
    //MARK: UnitSkillRateParameter
    /*****************************************************************************/

    ctor: function( /**@type {Element} */ xmlNode, /**@type {EU.Unit} */ unit )
    {
        this._rate = 1 ;
        this._parameter = "";

        this._super(xmlNode, unit);
    
        this._rate = EU.Common.strToFloat(EU.asObject(xmlNode.getAttribute( "rate" ), "1.0"));
        this._parameter = xmlNode.getAttribute( "parameter" );
        return true;
    },
    
    update: function( /**@type {var} */ dt, /**@type {EU.Unit} */ context ) {
        "use strict";

    }

});


